src/main/java/com/commercetools/sync/shoppinglists/utils/LineItemUpdateActionUtils.java

```diff
@@ -0,0 +1,257 @@
+            }
+
+            return newShoppingList.getLineItems()
+                                  .stream()
+                                  .filter(Objects::nonNull)
+                                  .map(AddLineItemWithSku::of)
+                                  .collect(toList());
+        }
+
+        final List<LineItem> oldLineItems = oldShoppingList.getLineItems();
+        final List<LineItemDraft> newlineItems = newShoppingList.getLineItems()
+                                                                .stream()
+                                                                .filter(Objects::nonNull)
+                                                                .collect(toList());
+
+        return buildUpdateActions(oldShoppingList, newShoppingList, oldLineItems, newlineItems, syncOptions);
+    }
+
+
+    /**
+     * The decisions in the calculating update actions are documented on the
+     * `docs/adr/0002-shopping-lists-lineitem-and-textlineitem-update-actions.md`
+     */
+    @Nonnull
+    private static List<UpdateAction<ShoppingList>> buildUpdateActions(
+        @Nonnull final ShoppingList oldShoppingList,
+        @Nonnull final ShoppingListDraft newShoppingList,
+        @Nonnull final List<LineItem> oldLineItems,
+        @Nonnull final List<LineItemDraft> newlineItems,
+        @Nonnull final ShoppingListSyncOptions syncOptions) {
+
+        final List<UpdateAction<ShoppingList>> updateActions = new ArrayList<>();
+
+        final int minSize = Math.min(oldLineItems.size(), newlineItems.size());
+        int indexOfFirstDifference = minSize;
+        for (int i = 0; i < minSize; i++) {
+
+            final LineItem oldLineItem = oldLineItems.get(i);
+            final LineItemDraft newLineItem = newlineItems.get(i);
+
+            if (oldLineItem.getVariant() == null || StringUtils.isBlank(oldLineItem.getVariant().getSku())) {
+
+                throw new IllegalArgumentException(
+                    format("LineItem at position '%d' of the ShoppingList with key '%s' has no SKU set. "
+                        + "Please make sure all line items have SKUs", i, oldShoppingList.getKey()));
+
+            } else if (StringUtils.isBlank(newLineItem.getSku())) {
+
+                throw new IllegalArgumentException(
+                    format("LineItemDraft at position '%d' of the ShoppingListDraft with key '%s' has no SKU set. "
+                        + "Please make sure all line items have SKUs", i, newShoppingList.getKey()));
+
+            }
+
+            if (oldLineItem.getVariant().getSku().equals(newLineItem.getSku())
+                && hasIdenticalAddedAtValues(oldLineItem, newLineItem)) {
+                // same sku, calculate actions.
+                updateActions.addAll(buildLineItemUpdateActions(
+                    oldShoppingList, newShoppingList, oldLineItem, newLineItem, syncOptions));
+            } else {
+                // different sku means the order is different.
```

Comment:
Or order is the same but addedAt value is different?

Reply:
But why do we have to do the update actions because of `addedAt`? You cannot update it according to the platform, so we should not care here IF we are not going to preserve it (which is very complex).

Reply:
I could update this comment..

In sync, it's different projects (or external resources) so not the same purpose. Here maybe you have a front end and you want to order them correctly.  It's not complex but hard to do and can only be done with removing and adding back which requires order change too.

Reply:
But it's not possible to do it with the platform, right? Why should we support it then? 
And even if you want to support it like you have it here, `addedAt` will not work if you change the order of line items by removing and adding. It will be overwritten anyway.

Reply:
I am not sure about this, maybe you are right we should not support because it's not supported as an action in API, @butenkor can you also check that?

It only makes sense for the front end which is ordering with `addedAt` values.

Reply:
TextLineItem is also not supporting addedAt changes, looks like forcing a change in here would not be necessary until we have a valid requirement from users of the library. I haven't implemented it for the text line items, and planning to remove it from line item actions too.